### PR TITLE
Feature/vsl integration

### LINKED ISSUE #194: Incorporate valuable changes from VSL's code
Gertjan provided a draft for improvements regarding his tasks [here](https://github.com/gertjan123/Met4FoF-redundancy). Lets review his changes and take everything that is useful for us especially from his [metrological_streams_v2.py](https://github.com/gertjan123/Met4FoF-redundancy/blob/master/Met4FoF_redundancy/agentMFred/metrological_streams_v2.py ).

### LINKED ISSUE #239: Introduce an amplitude parameter in generator classes
Currently, all the generator classes output signals with an amplitude 1. An extra parameter to set the
amplitude would be useful.

### LINKED ISSUE #247: Align cosine and sine default frequencies
At the moment the [CosineGenerator has a default frequency of 5](https://github.com/Met4FoF/agentMET4FOF/blob/99a1b0468574ed9ab348fc963457499c4b3c03eb/agentMET4FOF/streams.py#L416) and the [SineGenerator of 50](https://github.com/Met4FoF/agentMET4FOF/blob/develop/agentMET4FOF/streams.py#L388). That seems a bit arbitrary. Can we find a common best default?

### PR DESCRIPTION
**Pull request by anupam-prasad**
This mainly resolves #194.

A new module containing exceptions has been added. 

Check out the current version of the docs [here](https://agentmet4fof.readthedocs.io/en/feature-vsl_integration/).

Additionally it resolves #239:

`amplitude` has been replaced with `value` wherever necessary. The generator classes have a phase and amplitude parameter now.

Fixes #247 as well.

### CODE REVIEW

File: agentMET4FOF/metrological_agents.py [hunk 1]
```diff
@@ -356,4 +353,833 @@ def agent_loop(self):
...
+        input_data_maxlen: int
+
+        output_data_maxlen: int
+
+        """
+
+        if sensor_key_list is None:
+            sensor_key_list = []
+        super().init_parameters(input_data_maxlen=25, output_data_maxlen=25)
+        self.metadata = MetaData(
```
Comment (BjoernLudwigPTB) [inline agentMET4FOF/metrological_agents.py:400]: I would suggest to add type hints and declare those instance variables at the top of the class with for instance:

```python
metadata: MetaData
calc_type: str
[...] 
```

File: agentMET4FOF/metrological_agents.py [hunk 2]
```diff
@@ -356,4 +353,833 @@ def agent_loop(self):
...
+                    data = np.array([t_data_arr2d[-1, 0], ut_data_arr2d[-1, 0], ybest[0], uybest[0]])
+
+            # Send the data
+            if len(data.shape) == 1:
+                data = data.reshape((1, len(data)))
+
+            self.set_output_data(channel="default", data=data)
+            super().agent_loop()
+
+    def calc_consistent_estimates_no_corr(y_arr2d, uy_arr2d, prob_lim):
```
Comment (BjoernLudwigPTB) [inline agentMET4FOF/metrological_agents.py:546]: This is a method, so the first parameter needs to be `self` or we have to make it a static function.

**Review by BjoernLudwigPTB — COMMENTED** [review]
I had a first glance and already made some remarks. We should carefully go through this once in a code review meeting before we merge this and definitely include a working example/test to see if it runs through cleanly.

File: agentMET4FOF/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    """
+    Custom exception to handle the case when sensor results are not linearly independent
+    """
+    pass
+
+
+class SystemMatrixNotReducibleError(Exception):
+    """
+    Custom exception to handle the case when the system matrix *A* is not reducible
+    """
```
[the remark this answers is not in the document]
Comment (anupam-prasad) [reply agentMET4FOF/exceptions.py:9]: Looks good. I think "non-reducible" sounds better than "not reducible" though.

**Comment by BjoernLudwigPTB** [comment]
Commit 9e5ba36c might be of special interest to you. It seems, as if there were two bugs present. Maybe now the examples run?

**Comment by anupam-prasad** [comment]
> @anupam-prasad Commit [9e5ba36](https://github.com/Met4FoF/agentMET4FOF/commit/9e5ba36cfc26cf2b0041a4315d60115f2c598c61) might be of special interest to you. It seems, as if there were two bugs present. Maybe now the examples run?

That's strange. It certainly looks like a syntax error but it didn't raise any relevant errors.

**Comment by BjoernLudwigPTB** [comment]
@anupam-prasad Brilliant, does that mean, we are finished with this? What do you say?

**Comment by anupam-prasad** [comment]
> @anupam-prasad Brilliant, does that mean, we are finished with this? What do you say?

I think so. Ill take a look at Gertjan"s code once more to check if I missed anything. Will let you know tomorrow.

**Comment by anupam-prasad** [comment]
> @anupam-prasad Brilliant, does that mean, we are finished with this? What do you say?

I just took at Gertjan's code. The only thing that needs to be copied over is the documentation.

**Comment by BjoernLudwigPTB** [comment]
> > @anupam-prasad Brilliant, does that mean, we are finished with this? What do you say?
> 
> I just took at Gertjan's code. The only thing that needs to be copied over is the documentation.

Which docs do you refer to? Can you send a link?

**Comment by anupam-prasad** [comment]
> > > @anupam-prasad Brilliant, does that mean, we are finished with this? What do you say?
> > 
> > 
> > I just took at Gertjan's code. The only thing that needs to be copied over is the documentation.
> 
> Which docs do you refer to? Can you send a link?

There is some stuff [here](https://github.com/gertjan123/Met4FoF-redundancy/tree/master/docs)

**Comment by BjoernLudwigPTB** [comment]
> > > > @anupam-prasad Brilliant, does that mean, we are finished with this? What do you say?
> > > 
> > > 
> > > I just took at Gertjan's code. The only thing that needs to be copied over is the documentation.
> > 
> > 
> > Which docs do you refer to? Can you send a link?
> 
> There is some stuff [here](https://github.com/gertjan123/Met4FoF-redundancy/tree/master/docs)

I'll take care of that! Thanks for the heads up.

**Pull request merged**